lit_gpt/lora.py

```diff
@@ -271,13 +271,42 @@ def __init__(
             # ________________________________________
             # | query         | key       | value    |
             # ----------------------------------------
+            # For Llama2's GQA support, Q, K, and V weights are interleaved, so that weights for grouped
+            # queries are adjacent to their associated key and value weights.
+            # For example, suppose we have n_head = 12 with 3 query groups.
+            # Then along the embedding dimension the interleaved weights would look like
+            #
+            # [Q, Q, Q, Q, K, V, Q, Q, Q, Q, K, V, Q, Q, Q, Q, K, V],
+            #
+            # where each Q, K, and V has size head_dim.
+            #
+            # In this case, the previously-described weight update applies separately to each
+            # individual block, so the update will take the form
+            #
+            # [[ΔW,ΔW,ΔW, ..., 0,0,0, ..., ΔW,ΔW,ΔW, ΔW,ΔW,ΔW, ..., 0,0,0, ..., ΔW,ΔW,ΔW, ...],
+            #  [.............................................................................],
+            #  [ΔW,ΔW,ΔW, ..., 0,0,0, ..., ΔW,ΔW,ΔW, ΔW,ΔW,ΔW, ..., 0,0,0, ..., ΔW,ΔW,ΔW, ...]]
+            #      ↑              ↑            ↑        ↑             ↑            ↑
+            # ________________________________________________________________________________
+            # | q block 1 | k block 1  | v block 1 | q block 2 |  k block 2 |  v block 2 | ...
+            # --------------------------------------------------------------------------------
+            # Note that in the above diagram, the size of each q block will equal q_per_kv
+            # times the size of each k and v block.
+
+            q_per_kv = self.n_head // self.n_query_groups
+            slice_size = q_per_kv + 2
+            head_dim = out_features // (num_kv_heads * slice_size)
+            ind = range(out_features)
             self.lora_ind = []
             if enable_q:
-                self.lora_ind.extend(range(0, self.linear.in_features))
+                q_ind = list(filter(lambda x: (x // head_dim) % slice_size < slice_size - 2, ind))
+                self.lora_ind.extend(q_ind)
             if enable_k:
-                self.lora_ind.extend(range(self.linear.in_features, self.linear.in_features + self.kv_embd_size))
+                k_ind = list(filter(lambda x: (x // head_dim) % slice_size == slice_size - 2, ind))
+                self.lora_ind.extend(k_ind)
             if enable_v:
-                self.lora_ind.extend(range(self.linear.in_features + self.kv_embd_size, self.linear.out_features))
+                v_ind = list(filter(lambda x: (x // head_dim) % slice_size == slice_size - 1, ind))
+                self.lora_ind.extend(v_ind)
```

Comment:
Final indices must be in a sorted order.
```python
self.lora_ind.sort()
```

The reason being is that in the `zero_pad` method we use `torch.Tensor.index_copy` and, according to the [docs](https://pytorch.org/docs/stable/generated/torch.Tensor.index_copy_.html), the values will be copied in the exact order that is specified by provided indices.
In other words, when we want to pad a matrix, which structure is (q, q, q, q, k, v, ...) and first 6+ indices are for query matrix, then k and v will be copied to where queries are stored.

Reply:
I didn't understand this. Can you show your point in a bit of code?

Reply:
Sure, but it turns out that it's not necessary 🙂.
As I understand, there is no need for LoRA weights to repeat the same interleaved placement as in the pretrained weights. If we do not sort indices, the final LoRA weights will have structure [Q1, Q2, K1, K2, V1, V2] (if it's applied to all layers) and the correct placement will be done via `zero_pad` method.

Here are the code examples (LoRA is applied to Q and V matrices).

1. If we don't sort indices and LoRA has no interleaved placement of weights - everything is fine.

```python
import torch

n_embd = 8
n_head = 4

padded = torch.zeros((n_embd * 3), dtype=int)
print(f"{padded=}")
>> padded=tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

# NOTE: 1 - query updates, 3 - value updates
lora = torch.tensor([1, 1] * n_head + [3, 3] * n_head)
print(f"{lora=}")
>> lora=tensor([1, 1, 1, 1, 1, 1, 1, 1, 3, 3, 3, 3, 3, 3, 3, 3])

indices = [0, 1, 6, 7, 12, 13, 18, 19, 4, 5, 10, 11, 16, 17, 22, 23]
indices = torch.tensor(indices)

padded.index_copy(0, indices, lora)
>> tensor([1, 1, 0, 0, 3, 3, 1, 1, 0, 0, 3, 3, 1, 1, 0, 0, 3, 3, 1, 1, 0, 0, 3, 3])
```

2. If the indices are unsorted, but LoRA weights are interleaved, then the final order isn't correct.

```python
lora = torch.tensor([1, 1, 3, 3] * n_head)
print(f"{lora=}")
>> lora=tensor([1, 1, 3, 3, 1, 1, 3, 3, 1, 1, 3, 3, 1, 1, 3, 3])

indices = [0, 1, 6, 7, 12, 13, 18, 19, 4, 5, 10, 11, 16, 17, 22, 23]
indices = torch.tensor(indices)

padded.index_copy(0, indices, lora)
>> tensor([1, 1, 0, 0, 1, 1, 3, 3, 0, 0, 3, 3, 1, 1, 0, 0, 1, 1, 3, 3, 0, 0, 3, 3])
```

3, With sorted indices and interleaved LoRA weights, everything is fine.

```python
lora = torch.tensor([1, 1, 3, 3] * n_head)
print(f"{lora=}")
>> lora=tensor([1, 1, 3, 3, 1, 1, 3, 3, 1, 1, 3, 3, 1, 1, 3, 3])

indices = [0, 1, 6, 7, 12, 13, 18, 19, 4, 5, 10, 11, 16, 17, 22, 23]
indices.sort()
indices = torch.tensor(indices)

padded.index_copy(0, indices, lora)
>> tensor([1, 1, 0, 0, 3, 3, 1, 1, 0, 0, 3, 3, 1, 1, 0, 0, 3, 3, 1, 1, 0, 0, 3, 3])
```

----

I believe that technically all three cases might work, though I think that in order to learn, the model will converge to either the first or the third case. Since the third case requires sorting, and it's unnecessary, I think I have to admit that **I was wrong and there is no need to sort indices**.

Reply:
Thanks for the detailed analysis, in this case I will go with option (1) to avoid the sort.